Wrap output of LeakTrace (#1811)

Amokrane opened the pull request:
Text descriptions in the LeakTrace can be pretty long. This makes them hard to read on StackOverflow.Wrapping these at 80 characters which is the max width offered by StackOverflow code blocks.

![Screenshot 2020-05-03 at 01 53 22](https://user-images.githubusercontent.com/90436/80895221-cbec9b00-8ce2-11ea-8c3d-78575169aca7.png)

Amokrane commented:
Sort of naive implementation. Not sure whether the expectation here was to wrap after a word (here it breaks between leakcanar & y).

pyricau commented:
Thanks! I should probably have thought a bit more about this ahead of time.

I noticed this change wraps certain parts of the heap analysis string output, but not all. E.g. for a library leak the description can be quite long (the description is what prompted me to file this). Have you thought on running the wrapping code on the entire string output instead? Any reason not to do that?

The entry points `HeapAnalysisSuccess.toString()` & `HeapAnalysisError.toString()`.

The other thing I was thinking about is whether we should change the toString(), or provide another method (e.g. `toWrappedString()`) that takes a `width` parameter with a default value of 80, and then update callsites inside LeakCanary to use that. The result is that anyone currently using toString will not have any breaking change.

Thoughts?

Amokrane commented:
You are right, I should have wrapped the entire heap analysis string output not just certain parts of the LeakTrace.  I like the approach of offering a separate ```toWrappedString()``` method.

I will rework this.

Thanks for your review :)

Amokrane commented:
I looked into wrapping the entire string output and stumbled upon a few interesting edge cases. Let's take this output as an example:

```

┬───
│ GC Root: Local variable in native code
│
├─ android.net.ConnectivityThread instance
│    Leaking: NO (PathClassLoader↓ is not leaking)
│    Thread name: 'ConnectivityThread'
│    ↓ ConnectivityThread.contextClassLoader
├─ dalvik.system.PathClassLoader instance
│    Leaking: NO (CollaborateAdapter↓ is not leaking and A ClassLoader is never leaking)
│    ↓ PathClassLoader.runtimeInternalObjects
├─ java.lang.Object[] array
│    Leaking: NO (CollaborateAdapter↓ is not leaking)
│    ↓ Object[].[1559]
├─ com.android.rangroceryshopping.collaborate.CollaborateAdapter class
│    Leaking: NO (a class is never leaking)
│    ↓ com.android.rangroceryshopping.collaborate.CollaborateAdapter class static CollaborateAdapter.mOnClickListener
│                                                                                                    ~~~~~~
╰→ com.android.rangroceryshopping.collaborate.CollaborateActivity instance
​     Leaking: YES (ObjectWatcher was watching this because com.android.rangroceryshopping.collaborate.CollaborateActivity received Activity#onDestroy() callback and Activity#mDestroyed is true)
​     key = 2ac23d62-e7b9-49c8-b881-b8f32592027a
​     watchDurationMillis = 5353
​     retainedDurationMillis = 351

METADATA

Build.VERSION.SDK_INT: 28
Build.MANUFACTURER: samsung
LeakCanary version: 2.2
App process name: com.android.rangroceryshopping
Analysis duration: 19112 ms

```
A simple implementation that goes over each line and wraps them while keeping the correct indentation, yields the following output:

```
​
​┬───
​│ GC Root: Local variable in native code
​│
​├─ android.net.ConnectivityThread instance
​│    Leaking: NO (PathClassLoader↓ is not leaking)
​│    Thread name: 'ConnectivityThread'
​│    ↓ ConnectivityThread.contextClassLoader
​├─ dalvik.system.PathClassLoader instance
​│    Leaking: NO (CollaborateAdapter↓ is not leaking and A ClassLoader
​      is never leaking)
​│    ↓ PathClassLoader.runtimeInternalObjects
​├─ java.lang.Object[] array
​│    Leaking: NO (CollaborateAdapter↓ is not leaking)
​│    ↓ Object[].[1559]
​├─ com.android.rangroceryshopping.collaborate.CollaborateAdapter class
​│    Leaking: NO (a class is never leaking)
​│    ↓ com.android.rangroceryshopping.collaborate.CollaborateAdapter c
​       lass static CollaborateAdapter.mOnClickListener
​│                                                                     
​                               ~~~~~~
​╰→ com.android.rangroceryshopping.collaborate.CollaborateActivity inst
​   ance
​​     Leaking: YES (ObjectWatcher was watching this because com.androi
​     d.rangroceryshopping.collaborate.CollaborateActivity received Act
​     ivity#onDestroy() callback and Activity#mDestroyed is true)
​​     key = 2ac23d62-e7b9-49c8-b881-b8f32592027a
​​     watchDurationMillis = 5353
​​     retainedDurationMillis = 351
​
​METADATA
​
​Build.VERSION.SDK_INT: 28
​Build.MANUFACTURER: samsung
​LeakCanary version: 2.2
​App process name: com.android.rangroceryshopping
​Analysis duration: 19112 ms​
```
The interesting points to notice are:
1. The ~ characters which underline the reference name looses track of the reference if the line containing the reference breaks. That could be fixed by using a marker that is inlined instead of one that is on a separate line (for instance ->theReferenceName<-).

2. I suppose that lines starting with | should also start with | when they break (contrary to the output here, but that's easily fixable), but what about the ones starting with: ╰→?

In other words, should we have:
```
╰→ com.android.rangroceryshopping.collaborate.CollaborateActivity inst
​╰→ ance
```
or:
```
╰→ com.android.rangroceryshopping.collaborate.CollaborateActivity inst
​        ance
```

pyricau commented:
I'm a little bit confused with this line:

```
│    ↓ com.android.rangroceryshopping.collaborate.CollaborateAdapter class static CollaborateAdapter.mOnClickListener
```

I don't think LeakCanary can output this? As for the ~~~, it also doesn't map what LeakCanary normally outputs (it should be the same size in chars as the field). To be clear, I'm not saying your feedback is incorrect, just want to make sure I get the examples right.

Those are all really good points. The naive approach here is to cut at exactly 80 and add the remaining content on a new line with no indent / prefix, which leads to the problems you've highlighted.

To be frank, I hadn't thought about this too much. Let's see if we can come up with what's "ideal". Some thoughts:

1) Ideally, we probably word wrapping instead of cutting words in the middle. Best is probably to cut at the prior space, though for the class + fields we might want to cut on "." or []
2) Content on the following line should likely be indented / prefix in a way that "makes sense". Yes, that gets very tricky. For this:

```
​├─ dalvik.system.PathClassLoader instance
​│    Leaking: NO (CollaborateAdapter↓ is not leaking and A ClassLoader
​      is never leaking)
```

maybe we want this instead?

```
​├─ dalvik.system.PathClassLoader instance
​│    Leaking: NO (CollaborateAdapter↓ is not leaking and A ClassLoader
​​│    is never leaking)
```

For the `╰→` case I think you had the right idea of shifting with the same indent but no extra `╰→`.

3) What to do about `~~~~` ? We used to have markers like you suggested, and people had a hard time seeing them. I've found `~~~~` to be very useful. What it means, however, is that the line wrapping algorithm has to become very smart. When finding a line with `~~~~` it needs to look at the previous lines to find which word was highlighted, and then figure out where to put the `~~~`. This get real fun because you might go on a different line..

```
├─ com.example.FooFooFooFooFooFooFoo instance
│    Leaking: UNKNOWN
│    ↓ FooFooFooFooFooFooFoo.barbarbarbarbarbarbarbar
│                            ~~~~~~~~~~~~~~~~~~~~~~~~
```

```
├─ com.example.FooFooFooFooFooFooFoo instance
│    Leaking: UNKNOWN
│    ↓ FooFooFooFooFooFooFoo.
│      barbarbarbarbarbarbarbar
│      ~~~~~~~~~~~~~~~~~~~~~~~~
```

Amokrane commented:
Ah sorry for the confusion regarding that line, I modified it manually to highlight the problem, should have used lorem ipsum or a real example.

Yep this is trickier than expected. Looks challenging and fun, I will happily sink my teeth into it :)

Amokrane commented:
Just wanted to give a quick update: I have made some good progress on this: word wrapping, proper handling of indentation and line decoration (the │). Still need to fix one last edge case (as you said handling the ~ marker is really tricky) and tidy up the code a bit and then I will submit a new patch set for review.

ckesc commented:
Please add option to disable word wrapping as well. Sometimes it's more convenient to see class names without breaks :)

Amokrane commented:
I could do that yes. I suppose that a good place for that would be in the Config object?

Note that the wrapping can only occur after a `.` within a class canonical name.

Example of output that I get with the current implementation:
```
┬───
│ GC Root: Local variable in native code
│
├─ android.net.ConnectivityThread instance
│    Leaking: NO (PathClassLoader↓ is not leaking)
│    Thread name: 'ConnectivityThread'
│    ↓ ConnectivityThread.contextClassLoader
├─ dalvik.system.PathClassLoader instance
│    Leaking: NO (CollaborateAdapter↓ is not leaking and A ClassLoader is 
│    never leaking)Leaking: NO (CollaborateAdapter↓ is not leaking and A 
│    ClassLoader is never leaking)Leaking: NO (CollaborateAdapter↓ is not 
│    leaking and A ClassLoader is never leaking) 
│    ↓ PathClassLoader.runtimeInternalObjects
├─ java.lang.Object[] array
│    Leaking: NO (CollaborateAdapter↓ is not leaking)
│    ↓ Object[].[1559]
├─ com.android.rangroceryshopping.collaborate.CollaborateAdapter class
│    Leaking: NO (a class is never leaking)
│    ↓ com.android.rangroceryshopping.collaborate.CollaborateAdapter class 
│    static CollaborateAdapter.mOnClickListener 
│                              ~~~~~~
╰→ com.android.rangroceryshopping.collaborate.CollaborateActivity 
   instancecom.android.rangroceryshopping.collaborate.CollaborateActivity 
   instancecom.android.rangroceryshopping.collaborate.CollaborateActivity 
   instance 
​     Leaking: YES (ObjectWatcher was watching this because com.android.
​     rangroceryshopping.collaborate.CollaborateActivity received 
​     Activity#onDestroy() callback and Activity#mDestroyed is true) 
​     key = 2ac23d62-e7b9-49c8-b881-b8f32592027a
​     watchDurationMillis = 5353
​     retainedDurationMillis = 351

METADATA

Build.VERSION.SDK_INT: 28
Build.MANUFACTURER: samsung
LeakCanary version: 2.2
App process name: com.android.rangroceryshopping
Analysis duration: 19112 ms
```

Amokrane commented:
I added an attribute in the config to enable/disable leak trace wrapping. Please, let me know whether the line wrapping code itself is clear enough or not. There were some many exceptions to account  for (as mentioned in this thread), that the implementation ended up not being trivial. I did my best to comment complex areas.

I am obviously willing to take any feedback to make it more concise/cleaner/clearer, if possible.

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/LeakCanary.kt:
```diff
@@ -169,7 +169,14 @@ object LeakCanary {
+     *
+     * Defaults to true.
+     */
+    val wrapLeakTraceWhenSharing: Boolean = true
```
I'm not convinced we need to increase the API surface of the config for this. I expect that the decision to wrap or not is use case based rather than an overall setting for all of LeakCanary.

From an earlier comment:

> provide another method (e.g. toWrappedString()) that takes a width parameter with a default value of 80, and then update callsites inside LeakCanary to use that. The result is that anyone currently using toString will not have any breaking change.

So, to summarize:

* The behavior does not change when calling `HeapAnalysisSuccess.toString()`, we still get no wrapping
* A new method appears: `toWrappedString(width = 80)`
* Current call sites of `HeapAnalysisSuccess.toString()` are updated to call toWrappedString instead.

What do you think?

Amokrane replied:
I added this config based on cketti's suggestion in one of the comments. I also thought it could provide a quick workaround  if the wrapping algorithm didn't work well on some unexpected cases. However, I understand that we should be careful when increasing the API surface of the config. I will remove it.

That being said, with this change, `HeapAnalysisSucess.toString()` implementation is not impacted. In fact, the wrapping is applied externally on two specific use cases: `share` and `share_stack_overflow` in `LeakScreen`.

The advantage of this, compared to adding a `toWrappedString` to `HeapAnalysisSuccess`,  is that this is presentation logic and I thought it made more sense that it stays closer to the presentation code rather than adding it deep down in `shark`.

What do you think? Do you see other call sites? Do you see advantages in having the wrapping logic belong to `shark` instead?

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakScreen.kt:
```diff
@@ -198,12 +199,19 @@ private fun View.onLeakTraceSelected(analysis: HeapAnalysisSuccess, heapAnalysis
 Analysis duration: ${analysis.analysisDurationMillis} ms"""
+
+    if (LeakCanary.config.wrapLeakTraceWhenSharing) {
+      leakStr = LeakTraceWordWrapper.wrap(leakStr, 80)
```
nit: avoid 3 letter variable names (e.g. str), use full nouns instead. 

Here, instead of using a var you can return the original vs the wrapped one.

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt:
```diff
@@ -0,0 +1,368 @@
+/**
+ * Performs word wrapping of leak traces.
+ */
+object LeakTraceWordWrapper {
```
`internal`

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt:
```diff
@@ -0,0 +1,368 @@
+  /**
+   * This implements a greedy wrapping algorithm.
+   *
+   * <p>
```
Nit: keep an empty line between the two but remove `<p>` to be consistent with javadoc style elsewhere in the repo.

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt:
```diff
@@ -0,0 +1,368 @@
+   * indentation.
+   */
+  fun wrap(
+    heapAnalysisStr: String,
```
str ;)

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt:
```diff
@@ -0,0 +1,368 @@
+    var wrappedTrace = ""
+    val lines = heapAnalysisStr.lines()
+
+    for (l in lines.indices) {
```
`l` => `currentLineIndex`

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt:
```diff
@@ -0,0 +1,368 @@
+    // used to accurately track the marker position
+    var totalRemovedSpaces = 0
+
+    val stack = Stack<List<String>>()
```
The JDK `Stack` class should be avoided. ArrayList and ArrayDeque are good alternatives.

Amokrane replied:
The `ArrayDeque` from kotlin-stdlib is still marked as experimental. Are we okay with that? Or do you advise using the JDK one?

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt:
```diff
@@ -0,0 +1,368 @@
+   * <p>
+   *
+   * Each line is split into words, using spaces and dots as separators. From there, a wrapped
+   * version of the line is built by adding one word after the other until the string doesn't fit
```
Instead of splitting the entire line by words and adding until the threshold is reached, have you considered walking back from the character at WIDTH until you find a space or period?

Amokrane replied:
Yep that's a good idea. I will try that.

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt:
```diff
@@ -0,0 +1,368 @@
+            totalRemovedSpaces
+        ) ?: relativeMarkerPos
+
+        if (++cursor == words.size) {
```
Change `cursor++` and move it out of the `if` block for improved readability.

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt:
```diff
@@ -0,0 +1,368 @@
+        if (++cursor == words.size) {
+          break
+        }
+        nextWord =
+          fetchNextWord(
+              words[cursor]
+          )
```
Can this be on one line?

pyricau commented on leakcanary-android-core/src/test/java/leakcanary/internal/activity/screen/WrappedLineTest.kt:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  @Test
+  fun testSplitAndKeep() {
```
nit: test names should never start with `test`, this is an artifact of Junit 3 which didn't have the `@Test` annotation (test methods had to start with `testX` to be detected back then.

Also, you can use backticks if you want spaces in test names for JVM tests.

Amokrane replied:
Good to know, thanks!

pyricau commented on leakcanary-android-core/src/main/java/leakcanary/internal/activity/screen/LeakTraceWordWrapper.kt:
```diff
@@ -0,0 +1,368 @@
+ */
+object LeakTraceWordWrapper {
+
+  private const val SPACE_CHARACTER = '\u0020'
+  private const val TILDE_CHARACTER = '\u007E'
+  private const val PERIOD_CHARACTER = '\u002E'
+  private const val BREAKING_CHARACTER_LENGTH = 1
+  const val BREAK_CHARACTER = '\n'
```
* Consider removing the `_CHARACTER` suffix (this is clear enough)
* `BREAK_CHARACTER` should be private or internal
* Rename `BREAK_CHARACTER` to `LINE_BREAK` ?

Amokrane commented:
Thank you Py for this first look. I will address the comments and push a new patch set soon.

pyricau commented:
Let me know when you want me to take another look ;)

Amokrane commented:
Yep, I was planning to push a new patch set this week-end :)

Amokrane commented:
Quick update: rewrote and simplified the logic based on your feedback. Will clean up the code a bit and send a new patchset in the following days. (getting there...:P)

pyricau commented:
Thanks for the great work and for following up, sorry I didn't merge any earlier!

Amokrane commented:
No worries, I should have pinged you earlier as well. Thanks for the merge :)
